src/teleport.cpp

```diff
@@ -31,36 +31,42 @@ void Teleport::serializeAttr(PropWriteStream& propWriteStream) const
 	propWriteStream.write<uint8_t>(destPos.z);
 }
 
-ReturnValue Teleport::queryAdd(int32_t, const Thing&, uint32_t, uint32_t, Creature*) const
+ReturnValue Teleport::queryAdd(int32_t, std::shared_ptr<const Thing>, uint32_t, uint32_t,
+                               std::shared_ptr<Creature>) const
 {
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
-ReturnValue Teleport::queryMaxCount(int32_t, const Thing&, uint32_t, uint32_t&, uint32_t) const
+ReturnValue Teleport::queryMaxCount(int32_t, std::shared_ptr<const Thing>, uint32_t, uint32_t&, uint32_t) const
 {
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
-ReturnValue Teleport::queryRemove(const Thing&, uint32_t, uint32_t, Creature* /*= nullptr */) const
+ReturnValue Teleport::queryRemove(std::shared_ptr<const Thing>, uint32_t, uint32_t,
+                                  std::shared_ptr<Creature> /*= nullptr */) const
 {
 	return RETURNVALUE_NOERROR;
 }
 
-Cylinder* Teleport::queryDestination(int32_t&, const Thing&, Item**, uint32_t&) { return this; }
+std::shared_ptr<Thing> Teleport::queryDestination(int32_t&, std::shared_ptr<const Thing>, std::shared_ptr<Item>&,
+                                                  uint32_t&)
+{
+	return shared_from_this();
+}
 
-void Teleport::addThing(Thing* thing) { return addThing(0, thing); }
+void Teleport::addThing(std::shared_ptr<Thing> thing) { return addThing(0, thing); }
 
-void Teleport::addThing(int32_t, Thing* thing)
+void Teleport::addThing(int32_t, std::shared_ptr<Thing> thing)
 {
-	Tile* destTile = g_game.map.getTile(destPos);
+	auto destTile = g_game.map.getTile(destPos);
```

Comment:
Could this be `const auto`
```suggestion
	const auto destTile = g_game.map.getTile(destPos);
```